flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/scheduler/BatchSchedulingTest.java

```diff
@@ -0,0 +1,521 @@
+		// but no source has been configured.
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
+	}
+
+	@Test(expected = JobExecutionException.class)
+	public void testBatchSchedulingSourceWithInput() throws Exception {
+		// Create the JobGraph
+		JobGraph jobGraph = new JobGraph("testBatchSchedulingSourceWithInput");
+		jobGraph.setScheduleMode(ScheduleMode.BATCH_FROM_SOURCES);
+
+		AbstractJobVertex v1 = new AbstractJobVertex("v1");
+
+		AbstractJobVertex v2 = new AbstractJobVertex("v2");
+		v2.connectNewDataSetAsInput(v1, POINTWISE);
+
+		jobGraph.addVertex(v1);
+		jobGraph.addVertex(v2);
+
+		v2.setAsBatchSource();
+
+		// This should throw an Exception, because the configured source has an input vertex.
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
+	}
+
+	/**
+	 * <pre>
+	 *        O verify
+	 *        |
+	 *        . <------------- denotes a pipelined result
+	 *        O union
+	 *  +----´|`----+
+	 *  |     |     |
+	 *  ■     ■     ■ <------- denotes a blocking result
+	 *  O     O     O
+	 * src0  src1  src2
+	 * </pre>
+	 */
+	@Test
+	public void testBatchSchedulingLegWise() throws Exception {
+		// Create the JobGraph
+		JobGraph jobGraph = new JobGraph("testBatchSchedulingLegWise");
+		jobGraph.setScheduleMode(ScheduleMode.BATCH_FROM_SOURCES);
+
+		// Union
+		AbstractJobVertex union = new AbstractJobVertex("union");
+		union.setInvokableClass(UnionForwarder.class);
+		union.setParallelism(PARALLELISM);
+		jobGraph.addVertex(union);
+
+		// Create source vertices
+		AbstractJobVertex[] src = new AbstractJobVertex[3];
+		for (int i = 0; i < src.length; i++) {
+			src[i] = new AbstractJobVertex("src " + i);
+			src[i].setInvokableClass(SourceTask.class);
+			src[i].setParallelism(PARALLELISM);
+			src[i].getConfiguration().setInteger("index", i);
+			jobGraph.addVertex(src[i]);
+
+			// Connect sources to union node
+			union.connectNewDataSetAsInput(src[i], POINTWISE, BLOCKING);
+		}
+
+		// Create verify
+		AbstractJobVertex verify = new AbstractJobVertex("verify");
+		verify.setInvokableClass(VerifyIndexes.class);
+		verify.setParallelism(PARALLELISM);
+		verify.getConfiguration().setInteger("maxIndex", src.length);
+
+		verify.connectNewDataSetAsInput(union, POINTWISE, PIPELINED);
+		jobGraph.addVertex(verify);
+
+		// Slot sharing group for the last pipeline
+		SlotSharingGroup ssg = new SlotSharingGroup(union.getID(), verify.getID());
+		union.setSlotSharingGroup(ssg);
+		verify.setSlotSharingGroup(ssg);
+
+		// - Configure batch scheduling ------------------------------------------------------------
+
+		// src0 is the first to go...
+		src[0].setAsBatchSource();
+
+		src[0].addBatchSuccessors(src[1]); // src0 => src1
+
+		src[1].addBatchSuccessors(src[2]); // src1 => src2
+
+		src[2].addBatchSuccessors(union); // src2 => [union => verify]
+
+		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
```

Comment:
Of course, a completing test relies on the proper functioning of the code
you test. But usually it also tests that the right things happen a little
bit more fine-grained than by simply observing whether a job succeeds or
not. IMHO, this puts strong assumptions on the code and requires that many
parts work correctly. For example, the testing job would also finish, if
Flink ignores the batch scheduling completely and executes the job
pipelined. You can't tell from this test if this is not happening. What you
could do, for example, is to use some static fields to check that the
operators are executed one after the other.
On Jun 4, 2015 5:11 PM, "Ufuk Celebi" <notifications@github.com> wrote:

> In
> flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/scheduler/BatchSchedulingTest.java
> <https://github.com/apache/flink/pull/754#discussion_r31729748>:
>
> > +		SlotSharingGroup ssg = new SlotSharingGroup(union.getID(), verify.getID());
> > +		union.setSlotSharingGroup(ssg);
> > +		verify.setSlotSharingGroup(ssg);
> > +
> > +		// - Configure batch scheduling ------------------------------------------------------------
> > +
> > +		// src0 is the first to go...
> > +		src[0].setAsBatchSource();
> > +
> > +		src[0].addBatchSuccessors(src[1]); // src0 => src1
> > +
> > +		src[1].addBatchSuccessors(src[2]); // src1 => src2
> > +
> > +		src[2].addBatchSuccessors(union); // src2 => [union => verify]
> > +
> > +		flink.submitJobAndWait(jobGraph, false, TestingUtils.TESTING_DURATION());
>
> The exact order is not tested, but whether they execute one after the
> other. It will deadlock or throw a NoResourcesAvailable if this is not the
> case. It's an integration test (the naming is off)... unit testing this is
> somewhat hard. Is this better answering the question or am I still missing
> something? Doesn't a test always rely on the proper functioning of what it
> tries to test and then it just fails if it is not functioning properly? [image:
> :smile:]
>
> We can certainly add another test for the exact order. I think that will
> involve interaction with the JM actors etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/754/files#r31729748>.
>